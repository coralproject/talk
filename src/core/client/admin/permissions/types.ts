import { GQLUSER_ROLE_RL } from "coral-framework/schema";

export type PermissionMap<AbilityType extends string, ContextType> = {
  [abilityType in AbilityType]: {
    [role in GQLUSER_ROLE_RL]?: (context?: ContextType) => boolean;
  };
};
