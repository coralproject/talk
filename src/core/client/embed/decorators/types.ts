import pym from "pym.js";

export type CleanupCallback = () => void;
export type Decorator = (pym: pym.Parent) => CleanupCallback | void;
