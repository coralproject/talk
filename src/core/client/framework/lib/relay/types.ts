import { _RefType } from "react-relay";

export type FragmentKeys<T> = {
  [P in keyof T]: T[P] extends _RefType<any> | null ? P : never
}[keyof T];

export type FragmentKeysNoLocal<T> = Exclude<FragmentKeys<T>, "local">;
