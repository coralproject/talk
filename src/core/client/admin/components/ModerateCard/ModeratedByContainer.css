$moderated-by: var(--v2-colors-mono-100);
$moderated-by-username: var(--v2-colors-mono-900);

.moderatedBy {
  font-size: var(--v2-font-size-1);
  font-weight: var(--v2-font-weight-primary-semi-bold);
  font-family: var(--v2-font-family-primary);
  line-height: var(--v2-line-height-reset);
  color: $moderated-by;
  text-transform: uppercase;
}

.moderatedByUsername {
  font-size: var(--v2-font-size-2);
  font-weight: var(--v2-font-weight-secondary-bold);
  font-family: var(--v2-font-family-secondary);
  line-height: var(--v2-line-height-reset);
  color: $moderated-by-username;
  text-align: center;
  padding-top: 1px;
}
