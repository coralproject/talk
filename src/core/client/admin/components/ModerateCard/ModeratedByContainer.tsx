import { Localized } from "fluent-react/compat";
import React, { useCallback } from "react";
import { graphql } from "react-relay";

import { withFragmentContainer } from "coral-framework/lib/relay";
import { BaseButton } from "coral-ui/components/v2";

import { ModeratedByContainer_comment } from "coral-admin/__generated__/ModeratedByContainer_comment.graphql";
import { ModeratedByContainer_viewer } from "coral-admin/__generated__/ModeratedByContainer_viewer.graphql";

import styles from "./ModeratedByContainer.css";

interface Props {
  viewer: ModeratedByContainer_viewer;
  comment: ModeratedByContainer_comment;
  onUsernameClicked: (id?: string | null) => void;
}

const ModeratedByContainer: React.FunctionComponent<Props> = ({
  comment,
  viewer,
  onUsernameClicked,
}) => {
  let moderatedBy: React.ReactElement | null;
  let id: string | null = null;
  if (comment.statusHistory.edges.length === 0) {
    moderatedBy = null;
  } else if (comment.statusHistory.edges[0].node.moderator === null) {
    moderatedBy = (
      <Localized id="moderate-comment-moderatedBySystem">System</Localized>
    );
  } else if (viewer.id === comment.statusHistory.edges[0].node.moderator.id) {
    moderatedBy = viewer.username ? <>{viewer.username}</> : null;
  } else {
    moderatedBy = <>{comment.statusHistory.edges[0].node.moderator.username}</>;
    id = comment.statusHistory.edges[0].node.moderator.id;
  }

  if (!moderatedBy) {
    return null;
  }

  const onClick = useCallback(() => {
    onUsernameClicked(id);
  }, [onUsernameClicked, comment]);

  return (
    <BaseButton onClick={onClick}>
      <Localized id="moderate-comment-moderatedBy">
        <div className={styles.moderatedBy}>Moderated By</div>
      </Localized>
      <div className={styles.moderatedByUsername}>{moderatedBy}</div>
    </BaseButton>
  );
};

const enhanced = withFragmentContainer<Props>({
  comment: graphql`
    fragment ModeratedByContainer_comment on Comment {
      id
      statusLiveUpdated
      statusHistory(first: 1) {
        edges {
          node {
            moderator {
              id
              username
            }
          }
        }
      }
    }
  `,
  viewer: graphql`
    fragment ModeratedByContainer_viewer on User {
      id
      username
    }
  `,
})(ModeratedByContainer);

export default enhanced;
