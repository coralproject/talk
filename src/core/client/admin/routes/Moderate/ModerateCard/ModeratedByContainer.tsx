import { Localized } from "fluent-react/compat";
import React from "react";
import { graphql } from "react-relay";

import { ModeratedByContainer_comment } from "coral-admin/__generated__/ModeratedByContainer_comment.graphql";
import { ModeratedByContainer_viewer } from "coral-admin/__generated__/ModeratedByContainer_viewer.graphql";
import { withFragmentContainer } from "coral-framework/lib/relay";

import styles from "./ModeratedByContainer.css";

interface Props {
  viewer: ModeratedByContainer_viewer;
  comment: ModeratedByContainer_comment;
}

const ModeratedByContainer: React.FunctionComponent<Props> = ({
  comment,
  viewer,
}) => {
  let moderatedBy: React.ReactElement | null;
  if (!comment.statusLiveUpdated || comment.statusHistory.edges.length === 0) {
    moderatedBy = null;
  } else if (comment.statusHistory.edges[0].node.moderator === null) {
    moderatedBy = (
      <Localized id="moderate-comment-moderatedBySystem">System</Localized>
    );
  } else if (viewer.id === comment.statusHistory.edges[0].node.moderator.id) {
    moderatedBy = null;
  } else {
    moderatedBy = <>{comment.statusHistory.edges[0].node.moderator.username}</>;
  }

  if (!moderatedBy) {
    return null;
  }

  return (
    <div>
      <Localized id="moderate-comment-moderatedBy">
        <div className={styles.moderatedBy}>Moderated By</div>
      </Localized>
      <div className={styles.moderatedByUsername}>{moderatedBy}</div>
    </div>
  );
};

const enhanced = withFragmentContainer<Props>({
  comment: graphql`
    fragment ModeratedByContainer_comment on Comment {
      id
      statusLiveUpdated
      statusHistory(first: 1) {
        edges {
          node {
            moderator {
              id
              username
            }
          }
        }
      }
    }
  `,
  viewer: graphql`
    fragment ModeratedByContainer_viewer on User {
      id
    }
  `,
})(ModeratedByContainer);

export default enhanced;
