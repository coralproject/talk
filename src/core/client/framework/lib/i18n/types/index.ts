export { default as FluentShortNumber } from "./FluentShortNumber";
