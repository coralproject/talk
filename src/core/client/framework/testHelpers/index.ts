export {
  default as createRelayEnvironment,
  CreateRelayEnvironmentParams,
} from "./createRelayEnvironment";
export { default as createFluentBundle } from "./createFluentBundle";
export { default as createSinonStub } from "./createSinonStub";
export {
  default as removeFragmentRefs,
  NoFragmentRefs,
} from "./removeFragmentRefs";
export { default as createUUIDGenerator } from "./createUUIDGenerator";
export * from "./denormalize";
export { default as replaceHistoryLocation } from "./replaceHistoryLocation";
