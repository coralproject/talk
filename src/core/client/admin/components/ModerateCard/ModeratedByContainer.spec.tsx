import React from "react";
import { createRenderer } from "react-test-renderer/shallow";

import { removeFragmentRefs } from "coral-framework/testHelpers";
import { PropTypesOf } from "coral-framework/types";

import ModeratedByContainer from "./ModeratedByContainer";

const ModeratedByContainerN = removeFragmentRefs(ModeratedByContainer);

it("viewer's username shows on moderation cards moderated by viewer", () => {
  const props: PropTypesOf<typeof ModeratedByContainerN> = {
    comment: {
      id: "comment-id",
      statusLiveUpdated: false,
      statusHistory: {
        edges: [
          {
            moderator: {
              id: "viewer",
              username: "viewer",
            },
          },
        ],
      },
    },
    viewer: {
      id: "viewer",
      username: "viewer",
    },
    onUsernameClicked: (id?: string | null) => {
      return;
    },
  };

  const renderer = createRenderer();
  renderer.render(<ModeratedByContainerN {...props} />);
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
