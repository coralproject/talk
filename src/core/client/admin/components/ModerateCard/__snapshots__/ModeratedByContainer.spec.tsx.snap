// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`viewer's username shows on moderation cards moderated by viewer 1`] = `
<ForwardRef(Relay(ModeratedByContainer))
  comment={
    Object {
      "id": "comment-id",
      "statusHistory": Object {
        "edges": Array [
          Object {
            "moderator": Object {
              "id": "viewer",
              "username": "viewer",
            },
          },
        ],
      },
      "statusLiveUpdated": false,
    }
  }
  onUsernameClicked={[Function]}
  viewer={
    Object {
      "id": "viewer",
      "username": "viewer",
    }
  }
/>
`;
